### PR TITLE
drivers/stmpe811: fix interrupt polarity & interrupt generation

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

The Interrupt on the stmpe811 is generated on the falling edge.
By observing the rising edge, we will only get an event if the interrupt gets cleared.

Now that the interrupt of the driver works properly, we don't have to rely on polling.
So also update the test application to make use of this driver feature.

### Testing procedure

Flash `tests/driver_stmpe811` on the `stm32f429i-disco`/`stm32f429i-disc1`.
Touch events should generate an event on the terminal, should also be a bit more responsive than before.

### Issues/PRs references

found while reviewing https://github.com/RIOT-OS/RIOT/pull/14188#discussion_r455306419

### CODE REVIEW

**Comment by leandrolanzieri** [comment]
With this PR I only get one pair of `Pressed!` and `Released!` events. After that it stays locked. Moreover, the position reported seems to be wrong. Any ideas?

**PR**
```
Welcome to pyterm!
Type '/exit' to exit.
2020-07-16 09:40:10,749 # Pressed!
2020-07-16 09:40:10,753 # X: 2332, Y:8192
2020-07-16 09:40:10,984 # Pressed!
2020-07-16 09:40:10,986 # Released!

# no more events when pressing
```

**Master**
```
Welcome to pyterm!
Type '/exit' to exit.
2020-07-16 09:39:33,360 # Pressed!
2020-07-16 09:39:33,363 # X: 102, Y:31
2020-07-16 09:39:33,376 # X: 103, Y:32
2020-07-16 09:39:33,389 # X: 106, Y:37
2020-07-16 09:39:33,402 # X: 106, Y:39
2020-07-16 09:39:33,416 # X: 105, Y:40
2020-07-16 09:39:33,429 # X: 105, Y:41
2020-07-16 09:39:33,442 # X: 106, Y:41
2020-07-16 09:39:33,455 # X: 106, Y:42
2020-07-16 09:39:33,468 # X: 107, Y:43
2020-07-16 09:39:33,482 # X: 108, Y:44
2020-07-16 09:39:33,494 # X: 109, Y:48
2020-07-16 09:39:33,508 # X: 109, Y:48
2020-07-16 09:39:33,521 # X: 109, Y:49
2020-07-16 09:39:33,534 # X: 108, Y:48
2020-07-16 09:39:33,547 # X: 106, Y:41
2020-07-16 09:39:33,558 # Released!
2020-07-16 09:39:35,394 # Pressed!
2020-07-16 09:39:35,397 # X: 106, Y:41
2020-07-16 09:39:35,410 # X: 90, Y:41
2020-07-16 09:39:35,423 # X: 90, Y:43
2020-07-16 09:39:35,436 # X: 90, Y:43
2020-07-16 09:39:35,449 # X: 89, Y:41
2020-07-16 09:39:35,462 # X: 88, Y:41
2020-07-16 09:39:35,475 # X: 88, Y:41
2020-07-16 09:39:35,488 # X: 88, Y:41
2020-07-16 09:39:35,501 # X: 89, Y:41
2020-07-16 09:39:35,515 # X: 90, Y:41
2020-07-16 09:39:35,527 # X: 92, Y:44
2020-07-16 09:39:35,540 # X: 92, Y:44
2020-07-16 09:39:35,554 # X: 92, Y:43
2020-07-16 09:39:35,567 # X: 92, Y:42
2020-07-16 09:39:35,580 # X: 91, Y:42
2020-07-16 09:39:35,593 # X: 91, Y:42
2020-07-16 09:39:35,606 # X: 91, Y:42
2020-07-16 09:39:35,617 # Released!
2020-07-16 09:39:36,508 # Pressed!
2020-07-16 09:39:36,510 # X: 93, Y:43
2020-07-16 09:39:36,524 # X: 95, Y:42
2020-07-16 09:39:36,536 # X: 97, Y:47
2020-07-16 09:39:36,549 # X: 97, Y:49
2020-07-16 09:39:36,562 # X: 97, Y:49
2020-07-16 09:39:36,575 # X: 97, Y:49
2020-07-16 09:39:36,589 # X: 97, Y:48
2020-07-16 09:39:36,601 # X: 98, Y:49
2020-07-16 09:39:36,615 # X: 98, Y:49
2020-07-16 09:39:36,628 # X: 98, Y:49
2020-07-16 09:39:36,640 # X: 99, Y:51
2020-07-16 09:39:36,654 # X: 101, Y:55
2020-07-16 09:39:36,667 # X: 105, Y:60
2020-07-16 09:39:36,680 # X: 105, Y:61
2020-07-16 09:39:36,693 # X: 106, Y:61
2020-07-16 09:39:36,706 # X: 107, Y:66
2020-07-16 09:39:36,718 # Released!
```

**Comment by aabadie** [comment]
> With this PR I only get one pair of Pressed! and Released! events.

Same here. This PR is changing the initial behavior of the test application: before, positions were continuously reported when pressed, so one could move the finger while pressed and follow the positions. This is not possible anymore when using the mutex mechanism.

The other problem is that the touch position is not reported anymore when pressing again.

File: drivers/stmpe811/stmpe811.c [hunk 1]
```diff
@@ -162,7 +162,7 @@ int stmpe811_init(stmpe811_t *dev, const stmpe811_params_t * params, touch_event
 
     if ((dev->params.int_pin != GPIO_UNDEF) && cb) {
         DEBUG("[stmpe811] init: configuring touchscreen interrupt\n");
-        gpio_init_int(dev->params.int_pin, GPIO_IN, GPIO_RISING, cb, arg);
+        gpio_init_int(dev->params.int_pin, GPIO_IN, GPIO_FALLING, cb, arg);
```
Comment (aabadie) [inline drivers/stmpe811/stmpe811.c:165]: I would only buy this change here: indeed the default polarity is falling (0) in the interrupt control register.

**Comment by benpicco** [comment]
Huh that’s odd - it worked yesterday, but it was late 😅
The data sheet mentions an auto-hibernation mode (and not much more about it) but the way we configure the driver it should be able to wake up on touch and generate an interrupt - but it doesn’t (after a while of inactivity).

I thought this fixed it, but looks like i was wrong.

**Comment by benpicco** [comment]
I found some more:

 - we'll always have to clear the Interrupt Status in order to receive new Interrupts
 - if we reset the fifo, we will not get any new positions

For me the test is working now without polling - although the `Pressed`/`Released` messages are wrong.

File: drivers/stmpe811/stmpe811.c [hunk 2]
```diff
@@ -262,12 +262,12 @@ int stmpe811_read_touch_state(const stmpe811_t *dev, stmpe811_touch_state_t *sta
...
     }
 
+    _clear_interrupt_status(dev);
+
     if ((val & STMPE811_TSC_CTRL_STA)) {
-        _clear_interrupt_status(dev);
         *state = STMPE811_TOUCH_STATE_PRESSED;
     }
     else {
-        _reset_fifo(dev);
```
Comment (aabadie) [inline drivers/stmpe811/stmpe811.c:270]: This is required otherwise on the next press you will get the last pressed position first.

**Comment by benpicco** [comment]
We also have to enable the FiFo threshold interrupt.
With this, touch events are generated reliably and we don't have to clear the FiFo manually.

The "Pressed!" / "Released!" messages in the driver test are now correct again and swiping is detected correctly.

**Comment by aabadie** [comment]
I don't like enabling interrupt on FIFO thresdhold: the application is continuously interrupted, mimicking a polling read. And I think this is not what we want if we use the mutex lock/unlock in the application. I also noticed that in the current version the first position is always invalid.

The diff below works much better for me: it only enables edge interrupt type, so only press/release events and it reset the fifo in different places as well as cleaning up the application code (touch state is read first after unlocking, then the position is read only if the touch screen is pressed). The first position is always correct. Releasing and pressing in another position return the correct position.

```diff
diff --git a/drivers/stmpe811/stmpe811.c b/drivers/stmpe811/stmpe811.c
index a7af459c2d..6abb155a09 100644
--- a/drivers/stmpe811/stmpe811.c
+++ b/drivers/stmpe811/stmpe811.c
@@ -166,11 +166,11 @@ int stmpe811_init(stmpe811_t *dev, const stmpe811_params_t * params, touch_event
 
         /* Enable touchscreen interrupt */
         ret += i2c_write_reg(STMPE811_DEV_I2C, STMPE811_DEV_ADDR,
-                             STMPE811_INT_EN, STMPE811_INT_EN_TOUCH_DET | STMPE811_INT_EN_FIFO_TH, 0);
+                             STMPE811_INT_EN, STMPE811_INT_EN_TOUCH_DET, 0);
 
         /* Enable global interrupt */
         ret += i2c_write_reg(STMPE811_DEV_I2C, STMPE811_DEV_ADDR,
-                             STMPE811_INT_CTRL, STMPE811_INT_CTRL_GLOBAL_INT, 0);
+                             STMPE811_INT_CTRL, (STMPE811_INT_CTRL_GLOBAL_INT | STMPE811_INT_CTRL_INT_TYPE), 0);
     }
 
     if (ret < 0) {
@@ -193,6 +193,8 @@ int stmpe811_read_touch_position(stmpe811_t *dev, stmpe811_touch_position_t *pos
 {
     uint16_t tmp_x, tmp_y;
 
+    _reset_fifo(dev);
+
     /* Acquire I2C device */
     i2c_acquire(STMPE811_DEV_I2C);
 
@@ -200,7 +202,7 @@ int stmpe811_read_touch_position(stmpe811_t *dev, stmpe811_touch_position_t *pos
     uint32_t xyz_ul;
 
     if (i2c_read_regs(STMPE811_DEV_I2C, STMPE811_DEV_ADDR,
-                      STMPE811_TSC_DATA_NON_INC, &xyz, sizeof(xyz), 0) < 0) {
+                      STMPE811_TSC_DATA_XYZ, &xyz, sizeof(xyz), 0) < 0) {
         DEBUG("[stmpe811] position: cannot read position\n");
         i2c_release(STMPE811_DEV_I2C);
         return -STMPE811_ERR_I2C;
@@ -269,6 +271,7 @@ int stmpe811_read_touch_state(const stmpe811_t *dev, stmpe811_touch_state_t *sta
         *state = STMPE811_TOUCH_STATE_PRESSED;
     }
     else {
+        _reset_fifo(dev);
         *state = STMPE811_TOUCH_STATE_RELEASED;
     }
 
diff --git a/tests/driver_stmpe811/main.c b/tests/driver_stmpe811/main.c
index 12bdd60d1e..43d779b43e 100644
--- a/tests/driver_stmpe811/main.c
+++ b/tests/driver_stmpe811/main.c
@@ -51,14 +51,9 @@ int main(void)
     stmpe811_touch_state_t last_touch_state = current_touch_state;
 
     while (1) {
-        stmpe811_touch_position_t position;
-
         /* wait for touch event */
         mutex_lock(&lock);
-
         stmpe811_read_touch_state(&dev, &current_touch_state);
-        stmpe811_read_touch_position(&dev, &position);
-
         if (current_touch_state != last_touch_state) {
             if (current_touch_state == STMPE811_TOUCH_STATE_PRESSED) {
                 puts("Pressed!");
@@ -71,6 +66,8 @@ int main(void)
 
         /* Display touch position if pressed */
         if (current_touch_state == STMPE811_TOUCH_STATE_PRESSED) {
+            stmpe811_touch_position_t position;
+            stmpe811_read_touch_position(&dev, &position);
             printf("X: %i, Y:%i\n", position.x, position.y);
         }
     }
```

**Comment by aabadie** [comment]
@benpicco, are you ok with the diff proposed in https://github.com/RIOT-OS/RIOT/pull/14531#issuecomment-660941670 ? It makes the situation much better IMHO.

**Comment by benpicco** [comment]
With that patch we only get a single event on touch and a single event on release, but no swipe events.

> I don't like enabling interrupt on FIFO threshold: the application is continuously interrupted, mimicking a polling read.

Only when the user is performing a touch gesture, but we could reduce the number of interrupts by setting the FIFO threshold higher and extend the API to report multiple events :wink: 

But if you prefer a very basic touch API we can go that route. Users might be surprised why they don't get any events when moving on the screen. I sure got excited and was wondering whether I could draw numbers and let them be recognized by `emlearn`.

**Comment by aabadie** [comment]
> Users might be surprised why they don't get any events when moving on the screen

There's no gesture detection in that device. I think that there's no need to post events at the FIFO buffering rate (or even reduced). We could implement that anyway as an extension of the driver API. For the moment, the API is limited so it makes no sense to implement a more clever behavior under the hood.
The current API still allows gesture (or swipe) detection: just poll the position while pressed. This is how it's handled by lvgl for example.

> I sure got excited and was wondering whether I could draw numbers and let them be recognized by emlearn

I played with that in the early 2020 using tensorflow-lite but I didn't require FIFO events to get the positions. Just the simple algorithm:
```
while (1)
  if pressed:
    read_position_to_buffer
  if released:
    detect_digit
    empty_buffer
  sleep(sample_rate)
```

With this PR, it could be adapted to something like:
```
while (1)
  mutex_lock()
  while (pressed)
    read_position_to_buffer
  detect_digit
  empty_buffer
```

**Comment by benpicco** [comment]
In `stmpe811_read_touch_position()` you do the `_reset_fifo()` *before* `i2c_acquire()`.
When I move it past it, it will discard the current touch position and we always read `X: 2756, Y:8192`.

When I remove it entirely it works - so I'll just do that :smile:

**Comment by aabadie** [comment]
> When I remove it entirely it still works - so I'll just do that

Ok ! I think I left like this after a hard debugging session :)

**Comment by benpicco** [comment]
Wait, no - if I remove it entirely it also only reads invalid positions. It *only* works if it's before the `i2c_acquire()` :fearful:

**Comment by aabadie** [comment]
> It only works if it's before the i2c_acquire()

If it's possible to disable the FIFO and read positions on-demand, things would be simpler. But I couldn't find how to do that.

**Comment by benpicco** [comment]
We still shouldn't have to rely on what is basically undefined behavior. The RIOT API states that `i2c_acquire()` must be called before `i2c_write_reg()`, but if we do that this stops working.

**Comment by aabadie** [comment]
> We still shouldn't have to rely on what is basically undefined behavior. The RIOT API states that i2c_acquire() must be called before i2c_write_reg(), but if we do that this stops working.

Indeed, this is an issue! Maybe we should add a busy loop to check there's at least one measure in the FIFO before reading the position. So that would be: clear fifo, wait until there's one measure, read measure.

**Comment by aabadie** [comment]
The following works for me:

```diff
diff --git a/drivers/stmpe811/stmpe811.c b/drivers/stmpe811/stmpe811.c
index 988c6384d8..0ae365ca0b 100644
--- a/drivers/stmpe811/stmpe811.c
+++ b/drivers/stmpe811/stmpe811.c
@@ -193,11 +193,17 @@ int stmpe811_read_touch_position(stmpe811_t *dev, stmpe811_touch_position_t *pos
 {
     uint16_t tmp_x, tmp_y;
 
-    _reset_fifo(dev);
-
     /* Acquire I2C device */
     i2c_acquire(STMPE811_DEV_I2C);
 
+    _reset_fifo(dev);
+
+    uint8_t fifo_size = 0;
+    do {
+        i2c_read_reg(STMPE811_DEV_I2C, STMPE811_DEV_ADDR,
+                     STMPE811_FIFO_SIZE, &fifo_size, 0);
+    } while (!fifo_size);
+
     uint8_t  xyz[4];
     uint32_t xyz_ul;
```

**Comment by aabadie** [comment]
Even better (less reset_fifo calls):

```diff
diff --git a/drivers/stmpe811/stmpe811.c b/drivers/stmpe811/stmpe811.c
index 988c6384d8..db766dd177 100644
--- a/drivers/stmpe811/stmpe811.c
+++ b/drivers/stmpe811/stmpe811.c
@@ -179,8 +179,6 @@ int stmpe811_init(stmpe811_t *dev, const stmpe811_params_t * params, touch_event
         return -STMPE811_ERR_I2C;
     }
 
-    _reset_fifo(dev);
-
     /* Release I2C device */
     i2c_release(STMPE811_DEV_I2C);
 
@@ -193,11 +191,17 @@ int stmpe811_read_touch_position(stmpe811_t *dev, stmpe811_touch_position_t *pos
 {
     uint16_t tmp_x, tmp_y;
 
-    _reset_fifo(dev);
-
     /* Acquire I2C device */
     i2c_acquire(STMPE811_DEV_I2C);
 
+    _reset_fifo(dev);
+
+    uint8_t fifo_size = 0;
+    do {
+        i2c_read_reg(STMPE811_DEV_I2C, STMPE811_DEV_ADDR,
+                     STMPE811_FIFO_SIZE, &fifo_size, 0);
+    } while (!fifo_size);
+
     uint8_t  xyz[4];
     uint32_t xyz_ul;
 
@@ -271,7 +275,6 @@ int stmpe811_read_touch_state(const stmpe811_t *dev, stmpe811_touch_state_t *sta
         *state = STMPE811_TOUCH_STATE_PRESSED;
     }
     else {
-        _reset_fifo(dev);
         *state = STMPE811_TOUCH_STATE_RELEASED;
     }
 

```

**Comment by aabadie** [comment]
@benpicco, do we agree on https://github.com/RIOT-OS/RIOT/pull/14531#issuecomment-661825947 ?

**Comment by benpicco** [comment]
There you go.
I don't particularly like it, but it will do.

File: drivers/stmpe811/stmpe811.c [hunk 3]
```diff
@@ -194,6 +194,15 @@ int stmpe811_read_touch_position(stmpe811_t *dev, stmpe811_touch_position_t *pos
     /* Acquire I2C device */
     i2c_acquire(STMPE811_DEV_I2C);
 
+    /* we don't use the fifo yet, discard it and get the next event */
+    _reset_fifo(dev);
```
Comment (aabadie) [inline drivers/stmpe811/stmpe811.c:198]: I tried the last version of this PR with #14188 and it doesn't work well. If I move this line to line 269 (just before `*state = STMPE811_TOUCH_STATE_RELEASED;`, then it works like a charm in both the driver test application and #14188. Would mind applying this change @benpicco ?

Comment (aabadie) [reply drivers/stmpe811/stmpe811.c]: I allowed myself to push the update once again in your branch. I hope this is fine ;)

**Pull request merged**